Build distributions action

jared321 opened the pull request:
Self-review tasks:

- [x] Careful review of all changes
- [x] There was already ample confirmation that the action fails on build failures
- [x] Intentionally break code so that a test will fail and alter action so that only `make-sdist` is running tests.  Confirm that `make-sdist` correctly detects and reports test failure.
- [x] Intentionally break code so that a test will fail and alter action so that only `build-wheels` is running tests.  Confirm that `build-wheels` correctly detects and reports test failure.
- [x] Spot check logs of new action to determine that all wheels are being built and tested in an expected way
- [x] Download artifacts, inspect contents (e.g., does it include external dependencies such as libraries), and manually install and test
      * Tested Python 3.9 to 3.13 on ARM64 macOS.  Compiled Cython code only depends on OS libraries according to `otool -L`
      * Tested Python 3.10 and 3.11 on RedHat Linux 8.10 with x86_64 architecture.  Compiled Cython code only depends on OS libraries according to `ldd`
      * Tested Python 3.10 on Ubuntu 22.04 with x86_64 architecture.  Compiled Cython code only depends on OS libraries according to `ldd`.
- [x] Ensure that `release.yml` has been altered so that it only runs on pushes to `main`
- [x] Confirm that all actions are passing

@mosesyhc to review.  Upload these to Test-PyPI as part of review or wait until next release?  The last wheels automatically built are available as artifacts at
https://github.com/bandframework/surmise/actions/runs/17049773899

jared321 committed:
(Issue #164) Try building source distribution

jared321 committed:
(Issue #164) Try saving source distribution as artifact.

jared321 committed:
(Issue #164) Try building wheels.

jared321 committed:
(Issue #164) Don't build with Python 3.8.

jared321 committed:
(Issue #164) We aren't testing Python 3.14 yet.

jared321 committed:
(Issue #164) Skip intermittent failure and 3.14.

jared321 committed:
(Issue #164) Windows+Python 3.13 is still experimental.

jared321 committed:
(Issue #164) Try archiving wheels

jared321 committed:
(Issue #164) Cleaning a bit.

jared321 committed:
(Issue #164) Try testing wheels.

jared321 committed:
(Issue #164) Test source distribution as well.

jared321 committed:
(Issue #164) More cleaning.

jared321 committed:
(Issue #164) Cleaning as part of PR review.

jared321 committed:
(Issue #164) Intentionally break test to see if sdist action fails.

jared321 committed:
(Issue #164) Leave test broken and see if wheel builds fail.

The source distribution build/test job did indeed cause action to fail as
expected.

jared321 committed:
(Issue #164) Revert test and release action.

The macOS 14 build wheel action did indeed cause the action to fail as expected.

jared321 committed:
(Issue #164) Only build wheels on push to main.

mosesyhc committed:
long_description_content_type should specify x-rst to comply with PyPI

mosesyhc commented:
MC:
- [x] Downloaded artifacts, inspect contents
    * Tested Python 3.11 and 3.12 on Windows 11 OS x64.  Compiled Cython code only depends on OS libraries / DLLs, shown by [lucasg/Dependencies](https://github.com/lucasg/Dependencies) tool.
    * Noted that Python 3.13 wheel is **not** built currently.
- [ ] Test upload sample artifacts (from Python 3.11 and source) to test.pypi.org/
    * `surmise-0.1.dev1+g6deb4cd0a-cp311-cp311-macosx_10_9_x86_64.whl`
    * `surmise-0.1.dev1+g6deb4cd0a-cp311-cp311-macosx_11_0_arm64.whl`
    * `surmise-0.1.dev1+g6deb4cd0a-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.manylinux_2_28_x86_64.whl`
    * `surmise-0.1.dev1+g6deb4cd0a-cp311-cp311-win_amd64.whl`
    * `surmise-0.1.dev1+g6deb4cd0a.tar.gz`
- During the checking of the wheels and source files with `twine`, the uploading tool for PyPA/PyPI, using `twine check dist/*`, the following errors are revealed:
  - The produced wheels contains an incorrect version.  I expect it to at least be `v0.3.1+dev`, instead it shows `0.1+dev`.
    - @jared321 Why is the built surmise version 0.1?  Is it respecting any particular autoversioning?
  - The wheels did not pass `twine check`.
    - [x] In `setup.py`, the `long_description_content_type` should specify `x-rst` instead of `rst`. see [#0c35687](https://github.com/bandframework/surmise/pull/165/commits/0c3568711f56bad4b93c2b04f9d3de2fb2f1015b). The documentation files are inspected and verified to successfully compile and render via `tox -e html`.
    - [ ] After fixing the `x-rst` error raised by `twine`, a new wheel is built with `tox` and checked with `twine check` again.  Now the following error shows up with PyPI rst rendering.  The same error shows for all newly built wheels and source:
```
    Checking dist/surmise-0.3.1.dev94+gcfcf17b7c.d20250819.tar.gz: FAILED
    ERROR    `long_description` has syntax errors in markup and would not be
             rendered on PyPI.
             line 7: Error: Error in "image" directive:
             no content permitted.
    
             .. image:: https://badge.fury.io/py/surmise.svg
    
                :target: https://badge.fury.io/py/surmise
```

jared321 committed:
(Issue #164) Shallow checkouts don't include tags.

setuptools-scm uses tags in the project's clone to automatically determine the
version.  It appears that the default shallow clone did not include tags.  This
fix was present in publish.yml, which suggests that release.yml will hopefully
build distributions with the correct version.

mosesyhc commented:
Putting a note here about setuptools version impacting the autoversioning in building the correct release version.  In another package [mosesyhc/LCGP](https://github.com/mosesyhc/lcgp) I needed to upgrade `setuptools` to `setuptools>=80` to correctly build the release version.

jared321 committed:
(Issue #164) Make source distributions with full clone as well.

The wheels were building with the expected version string in the GH action.

mosesyhc commented:
I have downloaded the newly produced wheels from the release action.  I can verify that `twine check` works for linux and Mac wheels.  `twine check` on Windows wheel still fails with the same error as before, complaining about the error in image directives inside `README.rst`.  Is there a configuration issue I should be aware of?

jared321 commented:
And that fails when you run `twine check` on a Windows machine?

It appears that the `file` command line tool can help in Linux and macOS.  The original `README.rst` does not indicate any issues
```
README.rst: ASCII text
```
I can also use `vi` to visualize different carriage returns (`:set list`).  I only see typical unix carriage returns.

I can also run `file` on the `METADATA` files in the different wheels.  For macOS and Linux
```
METADATA: Unicode text, UTF-8 text
```

While for Windows, I see
```
METADATA: Unicode text, UTF-8 text, with CRLF, CR line terminators
```
I also see the use of the two carriage returns in `vi`.

It seems like `build` is using both carriage returns when making the Windows wheels.

Did you try publishing them to Test-PyPI to see if PyPI knows how to deal with this automatically?

mosesyhc commented:
When I test upload to testPyPI, I get hit with another error complaining about non-release versions (which I should have expected):

```
INFO     Response from https://test.pypi.org/legacy/:
         400 Bad Request
INFO     <html>
          <head>
           <title>400 The use of local versions in <Version('0.3.1.dev115+g9c834532f')> is not allowed. See
         https://packaging.python.org/specifications/core-metadata for more information.</title>
          </head>
          <body>
           <h1>400 The use of local versions in <Version('0.3.1.dev115+g9c834532f')> is not allowed. See
         https://packaging.python.org/specifications/core-metadata for more information.</h1>
           The server could not comply with the request since it is either malformed or otherwise incorrect.<br/><br/>
         The use of local versions in &lt;Version(&#x27;0.3.1.dev115+g9c834532f&#x27;)&gt; is not allowed. See
         https://packaging.python.org/specifications/core-metadata for more information.


          </body>
         </html>
```

jared321 committed:
(Issue #164) Use unix carriage return when loading long desc. README.

See if loading this file is the source of mixed carriage returns in Windows
wheels.

jared321 committed:
(Issue #164) Try the last version of cibuildwheel to build surmise wheels.

jared321 committed:
(Issue #164) Try different Windows runner version.

jared321 committed:
(Issue #164) Try setting version manually for use with Test-PyPI.

Also, reset to windows-latest runner.  Using newer OS didn't help.

jared321 committed:
(Issue #164) See if wheel builds are due to cibuildwheel

jared321 committed:
(Issue #164) Try just building a wheel manually for windows.

mosesyhc committed:
testing without using codecs for rst files

mosesyhc commented:
See TestPyPI results:
https://test.pypi.org/project/surmise/0.3.1/

jared321 committed:
(Issue #164) Cleaning up.

Ran tox tasks locally and all passed.

jared321 committed:
(Issue #164) Remove remnants of manual version specification.

jared321 committed:
(Issue #164) Turn on versioning with setuptools-scm again.

jared321 committed:
(Issue #164) Finalize the release action again.

jared321 commented:
* Finished clean-up
* Source distribution and all wheels passing `twine check`
* Retested source distribution and macOS wheels as above
  * Also installed from source distribution & tested on all machines
* Reset `release.yml` so it no longer runs on pull requests to `main`
* Reviewed all changes again
* All actions passing again

@mosesyhc  I'm happy with what we have and I am ready for the next and hopefully last part of your review.

mosesyhc reviewed (approved):
Confirmed that all the required wheels and source distribution pass twine now.
Once set with the next release tag, the package will be ready for release.
